scripts/assemble.py

```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env python3
+
+"""
+Generate the monolithic doctest.h header file.
+
+Our original script, scripts/cmake/assemble_single_header.cmake,
+simply read doctest_fwd.h and doctest.cpp and pasted them in-order,
+with some defines sandwiched inbetween.
+
+This script attempts to be a bit more future-proof, by automatically
+scanning for any .h and .cpp files that could be included,
+(topologically) sorting them by #include order, then creating
+the final header file.
+"""
+
+# /// script
+# requires-python = ">=3.6"
+# ///
+
+
+from __future__ import annotations
+
+import re
+import sys
+from pathlib import Path
+from typing import Generator, NoReturn
+
+
+def main(args: list[str]) -> NoReturn:
+  """Script entry-point."""
+
+  if len(args) != 1:
+    print("Usage: scripts/assemble.py", file=sys.stderr)
+    sys.exit(1)
+
+  script = Path(__file__).resolve()
+  root   = script.parent.parent
+  incdir = root / "doctest" / "parts"
+  output = root / "doctest" / "doctest.h"
+
+  headers = set(incdir.rglob("*.h"))
+  sources = set(incdir.rglob("*.cpp"))
+
+  def extract_header(line: str) -> str | None:
+    """
+    Extract a header file name from a line of C code.
+
+    Assuming the input looks something like:
+
+      ```c
+      #include "foo.h"
+      #include <bar.h>
+      ```
+
+    This function will return "foo.h" and "bar.h" respectively
+    """
+
+    matches = re.findall(r'#include\s*["<]([^">]+)[">]', line)
+    if len(matches) == 0:
+      return None
+    if len(matches) == 1:
+      return matches[0]
+
+    reason = f"'{line}' has multiple includes"
+    raise RuntimeError(reason)
+
+
+  def process_file(file: Path, /, visited: set[Path]) -> Generator[str, None, None]:
+    """
+    Process a file, yielding lines of code with #include's scrubbed.
+
+    Assuming the file represents a C source file, iterates over each line,
+    yielding the contents. If the line is an #include which has NOT
+    yet been seen (as indicated by the `visitor` set), then the
+    contents of THAT file is recursively run through this same method.
+
+    If the file HAS been visited, then it is ignored and the line is not yielded.
+
+    The effect of this is a fusion of:
+      1. A topological sort by-header, and
+      2. Inlining header content, and
+      3. Scrubbing #include's to other doctest files
+    """
+
+    if file in visited:
+      return
+
+    visited.add(file)
+    with file.open("r") as fd:
+      content = fd.read()
+
+    for line in content.splitlines(keepends=False):
+      header = extract_header(line)
+      if header is not None:
+        if (incdir / header) in headers:
+          yield from process_file(incdir / header, visited=visited)
+        else:
+          yield line
```

Comment:
Can be combined into `if header is not None and (incdir / header) in headers:` to remove one `else`.